### DISCUSSION TITLE
Custom provider?

### DISCUSSION THREAD

**Discussion by dominikjeske** [ChilliCream/hotchocolate#4059] [Q&A]
I'm new to GraphQL topic and I'm wonder is there any support in hotchocolate to write provider to other data sources - my idea is to use power of GraphQL with tooling and other existing infrastructure my company data specific infrastructure.

**Comment by michaelstaib** [comment]
Yes, we have abstracted the query providers.

We have implemented a couple of drivers. You can have a look at the MongoDB driver or the Neo4J one.
https://github.com/ChilliCream/hotchocolate/tree/develop/src/HotChocolate/Neo4J
https://github.com/ChilliCream/hotchocolate/tree/develop/src/HotChocolate/MongoDb

**Comment by PascalSenn** [comment]
@dominikjeske  Depends on what you mean by provider. HotChocolate is not bound to any specific data source. You can use it as a layer over your business logic to expose a GraphQL API. 
You can just write normal C# classes and methods and HotChocolate will map it to graphql types, fields and resolvers. What you do in the resolvers is completely up to you. 

Only if you want to make use of HotChocolate.Data (Filtering, Sorting, Projections) you will need specific providers. 
These are advanced features and in most cases you do not need it. We implemented these providers for common datasources. For you domain layer, you probably not really need custom providers. 
If you are interested this is a good read:
https://chillicream.com/docs/hotchocolate/api-reference/extending-filtering
And reference implementation would be MongoDb, EfCore or Neo4J
EFCore: https://github.com/ChilliCream/hotchocolate/tree/develop/src/HotChocolate/Data/src/Data/Filters/Expressions
MongoDB: https://github.com/ChilliCream/hotchocolate/tree/develop/src/HotChocolate/MongoDb
Neo4J: https://github.com/ChilliCream/hotchocolate/tree/develop/src/HotChocolate/Neo4J
